### PR TITLE
[WIP] Feature/impressions counter task

### PR DESCRIPTION
**Pull request by mmelograno**
# JS SDK

## What did you accomplish?
- Added `ImpressionCounterTask` to send impressions count into Split Servers

## Extra Notes

### CODE REVIEW

**Review by EmilianoSanchez — COMMENTED** [review]
some comments shared via slack

File: src/metrics/index.js [hunk 1]
```diff
@@ -35,13 +36,17 @@ import {
...
 
 const log = logFactory('splitio-metrics');
+const IMPRESSIONS_REFRESH_COUNT = 300000; // 5 minutes
 
 const MetricsFactory = context => {
   let impressionsRetries = 0;
   const settings = context.get(context.constants.SETTINGS);
   const storage = context.get(context.constants.STORAGE);
+  const impressionsCounter = context.get(context.constants.IMPRESSIONS_COUNTER);
+  const shouldPushImpressionsCount = settings.sync.impressionsMode === OPTIMIZED && settings.mode === STANDALONE_MODE;
```
Comment (NicoZelaya) [inline src/metrics/index.js:49]: Maybe we can make it "mode !== CONSUMER_MODE" to prepare for a possibly future producer behavior.

File: src/metrics/index.js [hunk 2]
```diff
@@ -35,13 +36,17 @@ import {
   MySegmentsCollector,
   ClientCollector
 } from './Collectors';
+import { OPTIMIZED, STANDALONE_MODE } from '../utils/constants';
 
 const log = logFactory('splitio-metrics');
+const IMPRESSIONS_REFRESH_COUNT = 300000; // 5 minutes
```
Comment (NicoZelaya) [inline src/metrics/index.js:42]: IMPRESSIONS_COUNT_RATE ? The refresh word was very crappily used in the past here.

File: src/metrics/index.js [hunk 3]
```diff
@@ -99,8 +104,50 @@ const MetricsFactory = context => {
       .then(() => latencyTrackerStop());
   };
 
+  const pushImpressionsCount = () => {
+    const imprCounts = impressionsCounter.size();
+    if (!impressionsCounter || imprCounts === 0) return Promise.resolve();
+
+    const pf = [];
+    const impressionsCount = impressionsCounter.popAll();
+    log.info(`Pushing ${imprCounts} impressions count`);
```
Comment (NicoZelaya) [inline src/metrics/index.js:113]: `Pushing count of impressions for ${imprCounts} features`

File: src/metrics/index.js [hunk 4]
```diff
@@ -99,8 +104,50 @@ const MetricsFactory = context => {
...
+
+    return impressionsService(impressionsCountRequest(settings, {
+      body: JSON.stringify({ pf })
+    }))
+      .then(() => {
+        impressionsRetries = 0;
+      })
+      .catch(err => {
+        if (impressionsRetries) { // For now we retry only once.
+          log.warn(`Droping ${imprCounts} impressions count after retry. Reason ${err}.`);
```
Comment (NicoZelaya) [inline src/metrics/index.js:139]: `Droping count of impressions for ${imprCounts} features after retry. Reason ${err}.`

File: src/metrics/index.js [hunk 5]
```diff
@@ -99,8 +104,50 @@ const MetricsFactory = context => {
...
+          impressionsRetries = 0;
+        } else {
+          impressionsRetries++;
+          log.warn(`Failed to push ${imprCounts} impressions count, keeping data to retry on next iteration. Reason ${err}.`);
```
Comment (NicoZelaya) [inline src/metrics/index.js:143]: `Failed to push impressions count for ${imprCounts} features, keeping data...`

File: src/metrics/index.js [hunk 6]
```diff
@@ -99,8 +104,50 @@ const MetricsFactory = context => {
...
+
+    const keys = Object.keys(impressionsCount);
+    for (let i = 0; i < keys.length; i++) {
+      const splitted = keys[i].split('::');
+      if (splitted.length !== 2) continue;
+      const featureName = splitted[0];
+      const timeFrame = splitted[1];
+
+      const impressionsInTimeframe = {
+        t: featureName, // Test Name
```
Comment (NicoZelaya) [inline src/metrics/index.js:123]: use `f` to match the other jsons.

**Review by NicoZelaya — APPROVED** [review]

**Review by EmilianoSanchez — APPROVED** [review]

**Review by EmilianoSanchez — APPROVED** [review]
LGTM

**Pull request merged**